tools/codegen/api/types.py

```diff
@@ -40,15 +51,179 @@ def __str__(self) -> str:
+# corresponding to it, based on how we desugar to C++.  See also
+# CppSignatureGroup.
+@dataclass(frozen=True)
+class CppSignature:
+    # The schema this signature is derived from
+    func: FunctionSchema
+
+    # Enough information about the C++ types to generate a full
+    # C++ type signature for this signature.  I'm not too sure
+    # if these are the right representations, so for now this
+    # is intended to be more abstract.  Prefer using
+    # cpp_grouped_arguments() to access this
+    _cpp_grouped_arguments: Tuple[Union[CppArgument, CppTensorOptionsArguments, CppThisArgument], ...]
+    _cpp_returns_type: str
+
+    # WARNING: This is probably NOT what you want
+    #
+    # Return the flattened argument structure of this signature,
+    # discarding information about grouping.  If you are planning
+    # to translate these arguments into expressions on another
+    # API, you probably want cpp_grouped_arguments instead (which
+    # will handle TensorOptions translations correctly)
+    def cpp_ungrouped_arguments(self) -> Sequence[CppArgument]:
+        return [sub_a for a in self._cpp_grouped_arguments for sub_a in a.explicit_arguments()]
+
+    # Return the grouped argument structure of this signature.  This
+    # preserves high-level structure of the arguments so you may
+    # find it easier to do translations working with this
+    # representation.
+    def cpp_grouped_arguments(self) -> Sequence[Union[CppArgument, CppTensorOptionsArguments, CppThisArgument]]:
+        return self._cpp_grouped_arguments
+
+    # Render the C++ declaration for this signature
+    def decl(self) -> str:
+        cpp_args_str = ', '.join(map(str, self.cpp_ungrouped_arguments()))
+        return f"{self._cpp_returns_type} {cpp.name(self.func)}({cpp_args_str})"
+
+    # Render the C++ definition for this signature, not including
+    # the body (with curly braces)
+    def defn(self, prefix: str = "") -> str:
+        cpp_args_str = ', '.join(a.str_no_default() for a in self.cpp_ungrouped_arguments())
+        return f"{self._cpp_returns_type} {prefix}{cpp.name(self.func)}({cpp_args_str})"
+
+    # NB: This constructor knows how to disambiguate defaults when
+    # faithful is True.  Ideally this would live as an external process
+    # see https://github.com/pytorch/pytorch/pull/45666
+    @staticmethod
+    def _from_grouped_arguments(
+        func: FunctionSchema,
+        arguments: Sequence[Union[Argument, TensorOptionsArguments, ThisArgument]],
+        *,
+        faithful: bool
+    ) -> 'CppSignature':
+        if faithful:
+            # Immediately, manually do overload disambiguation, by
+            # dropping defaults from the faithful signature.  In
+            # principle, we should be able to do this at some later
+            # point in time with other overload disambiguation
+            cpp_grouped_arguments = tuple(
+                cpp.argument_faithful(a).no_default() for a in arguments
+            )
```

Comment:
This is a good point. So do it properly we need to teach codegen about conversions and then consider those when looking for ambiguity.

Reply:
It's complicated because removing the default from layout would remove the `ScalarType` ambiguity, but also makes it impossible to call with just a `c10::optional<ScalarType>`.

Two solutions I can think of that wouldn't remove any call signatures:
- make the `TensorOptions` constructor `explicit`. If you pass just a `ScalarType` it selects the faithful overload without ambiguity.
- detect implicit conversions and add new overloads for those types. So, here a new `ScalarType` overload would be created that just picks one operator and calls it unambiguously. This could be a more general solution, maybe even solving issues like #40287. However, would be far more compilcated.

For `TensorOptions`, we could also treat it as one argument and remove all the defaults like you do here.

Reply:
So maybe it is best to keep the default stripping here an orthogonal mechanism from disambiguation. The faithful cpp signature is not really intended for users, it's more a convenience for other users who want to be able to uniformly access the API. I will say that this is kind of displeasing though :(

To answer your questions directly:

> make the TensorOptions constructor explicit. If you pass just a ScalarType it selects the faithful overload without ambiguity.

This is not great, because the one of the UX affordances of TensorOptions is if you only provide a ScalarType, you can provide it by itself, without having to wrap it in TensorOptions.

> detect implicit conversions and add new overloads for those types. So, here a new ScalarType overload would be created that just picks one operator and calls it unambiguously.

Yes I agree this would be complicated.